File: src/agents/storageStrategist.ts
import { BaseAgent } from './baseAgent';
import { StorageAnalysis, StorageConditions, MarketPrice } from '../types';

// Default storage loss rates (percentage per day)
const DEFAULT_LOSS_RATES: Record<string, number> = {
  'wheat': 0.5,    // 0.5% loss per day
  'rice': 0.8,     // 0.8% loss per day
  'corn': 0.6,     // 0.6% loss per day
  'soybean': 0.7,  // 0.7% loss per day
};

// Default shelf life in days for different crops
const DEFAULT_SHELF_LIFE: Record<string, number> = {
  'wheat': 180,    // ~6 months
  'rice': 210,     // ~7 months
  'corn': 150,     // ~5 months
  'soybean': 120,  // ~4 months
};

export class StorageStrategistAgent extends BaseAgent<StorageAnalysis> {
  private crop: string;
  private initialQuantity: number;
  private currentStorage: StorageConditions;
  private marketPrices: MarketPrice[];

  constructor(
    crop: string,
    initialQuantity: number,
    currentStorage: Partial<StorageConditions> = {},
    marketPrices: MarketPrice[] = []
  ) {
    super('StorageStrategistAgent');
    this.crop = crop.toLowerCase();
    this.initialQuantity = initialQuantity;
    this.marketPrices = marketPrices;
    
    // Set default storage conditions with overrides
    this.currentStorage = {
      temperature: 25, // Celsius
      humidity: 60,    // percentage
      shelfLife: DEFAULT_SHELF_LIFE[this.crop] || 90, // days
      dailyLossRate: this.calculateAdjustedLossRate(currentStorage)
    };
  }

  private calculateAdjustedLossRate(storage: Partial<StorageConditions>): number {
    // Base loss rate from defaults
    let lossRate = DEFAULT_LOSS_RATES[this.crop] || 1.0; // Default to 1% if crop not found
    
    // Adjust based on storage conditions
    if (storage.temperature !== undefined) {
      // Ideal storage temperature is around 20°C for most grains
      const tempDiff = Math.abs(20 - storage.temperature);
      lossRate += tempDiff * 0.05; // 0.05% increase in loss per degree from ideal
    }
    
    if (storage.humidity !== undefined) {
      // Ideal humidity is around 60%
      const humidityDiff = Math.abs(60 - storage.humidity);
      lossRate += humidityDiff * 0.02; // 0.02% increase in loss per % from ideal
    }
    
    return Math.min(5, Math.max(0.1, lossRate)); // Keep between 0.1% and 5%
  }

  private calculateValueOverTime(
    days: number,
    initialValue: number,
    dailyLossRate: number
  ): { day: number; remainingValue: number; cumulativeLoss: number }[] {
    const result = [];
    let currentValue = initialValue;
    
    for (let day = 0; day <= days; day++) {
      if (day > 0) {
        // Apply daily loss
        const dailyLoss = currentValue * (dailyLossRate / 100);
        currentValue -= dailyLoss;
      }
      
      result.push({
        day,
        remainingValue: parseFloat((currentValue / initialValue * 100).toFixed(2)),
        cumulativeLoss: parseFloat((initialValue - currentValue).toFixed(2))
      });
    }
    
    return result;
  }

  async execute(daysToProject: number = 30) {
    const steps = [];
    
    try {
      // Calculate storage losses over time
      const projectedLosses = this.calculateValueOverTime(
        daysToProject,
        this.initialQuantity,
        this.currentStorage.dailyLossRate
      );
      
      steps.push(this.logStep('calculateLosses', { 
        days: daysToProject, 
        initialQuantity: this.initialQuantity,
        dailyLossRate: this.currentStorage.dailyLossRate
      }));

      // Calculate max storage days based on shelf life and value decay
      const maxStorageDays = Math.min(
        this.currentStorage.shelfLife,
        Math.ceil(100 / this.currentStorage.dailyLossRate) // Days until value drops to near zero
      );

      // Prepare response
      const response: StorageAnalysis = {
        currentStorage: this.currentStorage,
        projectedLosses,
        maxStorageDays
      };

      return this.createResponse(true, response, steps);
    } catch (error) {
      console.error('StorageStrategistAgent error:', error);
      steps.push(this.logStep('error', { error: error.message }));
      return this.createResponse(false, null as any, steps);
    }
  }
}

File: backend/src/agents/storageStrategist.py
from typing import List, Dict, Any
from datetime import datetime
from ..types import StorageAnalysis, StorageConditions

# Default storage loss rates (percentage per day)
DEFAULT_LOSS_RATES: Dict[str, float] = {
    'wheat': 0.5,
    'rice': 0.8,
    'corn': 0.6,
    'soybean': 0.7,
    'tomato': 3.5,
    'onion': 0.8,
    'potato': 0.5,
}

# Default shelf life in days
DEFAULT_SHELF_LIFE: Dict[str, int] = {
    'wheat': 180,
    'rice': 210,
    'corn': 150,
    'soybean': 120,
    'tomato': 14,
    'onion': 60,
    'potato': 90,
}

class StorageStrategistAgent:
    def __init__(self, crop: str, initial_quantity: float, current_storage: Dict[str, Any] = None):
        self.crop = crop.lower()
        self.initial_quantity = initial_quantity
        self.current_storage = current_storage or {}
        self.steps = []
    
    def _log_step(self, action: str, result: Any):
        step = {
            "action": action,
            "result": result,
            "timestamp": datetime.utcnow().isoformat()
        }
        self.steps.append(step)
        return step
    
    def _calculate_adjusted_loss_rate(self) -> float:
        """Calculate loss rate adjusted for storage conditions"""
        base_loss_rate = DEFAULT_LOSS_RATES.get(self.crop, 1.0)
        
        # Adjust based on storage conditions
        if 'temperature' in self.current_storage:
            temp_diff = abs(20 - self.current_storage['temperature'])
            base_loss_rate += temp_diff * 0.05
        
        if 'humidity' in self.current_storage:
            humidity_diff = abs(60 - self.current_storage['humidity'])
            base_loss_rate += humidity_diff * 0.02
        
        return min(5.0, max(0.1, base_loss_rate))
    
    def _calculate_value_over_time(self, days: int, initial_value: float, daily_loss_rate: float) -> List[Dict[str, Any]]:
        """Calculate value degradation over time"""
        result = []
        current_value = initial_value
        
        for day in range(days + 1):
            if day > 0:
                daily_loss = current_value * (daily_loss_rate / 100)
                current_value -= daily_loss
            
            result.append({
                "day": day,
                "remainingValue": round((current_value / initial_value * 100), 2),
                "cumulativeLoss": round((initial_value - current_value), 2)
            })
        
        return result
    
    async def execute(self, days_to_project: int = 30):
        """Execute storage analysis"""
        import asyncio
        
        try:
            # Calculate adjusted loss rate
            daily_loss_rate = self._calculate_adjusted_loss_rate()
            
            # Calculate storage losses over time
            projected_losses = self._calculate_value_over_time(
                days_to_project,
                self.initial_quantity,
                daily_loss_rate
            )
            
            self._log_step('calculateLosses', {
                'days': days_to_project,
                'initialQuantity': self.initial_quantity,
                'dailyLossRate': daily_loss_rate
            })
            
            # Calculate max storage days
            shelf_life = DEFAULT_SHELF_LIFE.get(self.crop, 90)
            max_storage_days = min(
                shelf_life,
                int(100 / daily_loss_rate) if daily_loss_rate > 0 else shelf_life
            )
            
            # Create storage conditions
            storage_conditions = StorageConditions(
                temperature=self.current_storage.get('temperature', 25),
                humidity=self.current_storage.get('humidity', 60),
                shelfLife=shelf_life,
                dailyLossRate=daily_loss_rate
            )
            
            # Prepare response
            analysis = StorageAnalysis(
                currentStorage=storage_conditions,
                projectedLosses=projected_losses,
                maxStorageDays=max_storage_days
            )
            
            
            return {
                "success": True,
                "data": analysis,
                "metadata": {
                    "timestamp": datetime.utcnow().isoformat(),
                    "processingTime": 0,
                    "agent": "StorageStrategistAgent"
                },
                "trace": {
                    "steps": self.steps
                }
            }
        except Exception as e:
            self._log_step('error', {'error': str(e)})
            return {
                "success": False,
                "data": None,
                "metadata": {
                    "timestamp": datetime.utcnow().isoformat(),
                    "processingTime": 0,
                    "agent": "StorageStrategistAgent"
                },
                "trace": {
                    "steps": self.steps
                }
            }

